src/main.cpp

```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
 {
     AssertLockHeld(cs_main);
     bool fInvalidFound = false;
     const CBlockIndex *pindexOldTip = chainActive.Tip();
     const CBlockIndex *pindexFork = chainActive.FindFork(pindexMostWork);
 
-    // Disconnect active blocks which are no longer in the best chain.
     bool fBlocksDisconnected = false;
+    boost::thread::id this_id(boost::this_thread::get_id()); // get this thread's id
     while (chainActive.Tip() && chainActive.Tip() != pindexFork) {
+        // When running in parallel block validation mode it is possible that this competing block could get to this
+        // point just after the chaintip had already been advanced.  If that were to happen then it could initiate a
+        // re-org when in fact a Quit had already been called on this thread.  So we do a check if Quit was previously
+        // called and return if true.
+        if (PV.QuitReceived(this_id))
+            return false;
+
+        // Disconnect active blocks which are no longer in the best chain.
         if (!DisconnectTip(state, chainparams.GetConsensus()))
             return false;
+
+        // Once the first block has been disconnected and a re-org has begun then we need to terminate any
+        // currently running PV threads that are validating.  They will likley have self terminated
+        // at this point anyway because the chain tip and UTXO base view will have changed but just
+        // to be sure we are not waiting on script threads to finish we can issue the termination here.
+        if (!fBlocksDisconnected) {
+            PV.StopAllValidationThreads(this_id);
+        }
         fBlocksDisconnected = true;
+
     }
 
     // Build list of new blocks to connect.
     std::vector<CBlockIndex*> vpindexToConnect;
     bool fContinue = true;
+    /** Parallel Validation: fBlock determines whether we pass a block or NULL to ConnectTip().
+     *  If the pindexMostWork has been extended while we have been validating the last block then we
+     *  want to pass a NULL so that the next block is read from disk, because we will definitely not 
+     *  have the block. 
+     */
+    bool fBlock = true;
     int nHeight = pindexFork ? pindexFork->nHeight : -1;
-    while (fContinue && nHeight != pindexMostWork->nHeight) {
+    while (fContinue && nHeight < pindexMostWork->nHeight) {
+
+        // During IBD if there are many blocks to connect still it could be a while before shutting down
+        // and the user may think the shutdown has hung, so return here and stop connecting any remaining
+        // blocks.
+        if (ShutdownRequested())
+            return false;
+
         // Don't iterate the entire list of potential improvements toward the best tip, as we likely only need
         // a few blocks along the way.
-        int nTargetHeight = std::min(nHeight + 32, pindexMostWork->nHeight);
+        int nTargetHeight = std::min(nHeight + (int)BLOCK_DOWNLOAD_WINDOW, pindexMostWork->nHeight);
```

Comment:
because block_download_window is unsigned and if we don't we get compiler warnings on some systems.